interface/super/layout_service_codes.php

```diff
@@ -0,0 +1,224 @@
+        if ($services) $services .= ';';
+        $services .= $key;
+      }
+      if (!$form_dryrun) {
+        sqlStatement("UPDATE layout_group_properties SET grp_services = ? WHERE " .
+          "grp_form_id = ? AND grp_group_id = ''",
+          array($services, $layoutid));
+      }
+    }
+  } // end upload logic
+}
+
+?>
+<form method='post' action='layout_service_codes.php' enctype='multipart/form-data'
+ onsubmit='return top.restoreSession()'>
+
+<center>
+
+<p class='text'>
+<table border='1' cellpadding='4'>
+ <tr bgcolor='#dddddd' class='dehead'>
+  <td align='center' colspan='2'>
+   <?php echo xlt('Install Layout Service Codes'); ?>
+  </td>
+ </tr>
+ <tr>
+  <td class='detail' nowrap>
+   <?php echo htmlspecialchars(xl('Source CSV File')); ?>
+   <input type="hidden" name="MAX_FILE_SIZE" value="350000000" />
+  </td>
+  <td class='detail' nowrap>
+   <input type="file" name="form_file" size="40" />
+  </td>
+ </tr>
+ <tr>
+  <td class='detail' nowrap>
+   <?php echo xlt('Test only, skip updates'); ?>
+  </td>
+  <td class='detail' >
+   <input type='checkbox' name='form_dryrun' value='1' checked />
+  </td>
+ </tr>
+ <tr bgcolor='#dddddd'>
+  <td align='center' class='detail' colspan='2'>
+   <input type='submit' name='bn_upload' value='<?php echo xlt('Upload and Apply') ?>' />
+  </td>
+ </tr>
+</table>
+</p>
+
+<p>
+<?php echo xlt('The input should be a CSV file with 3 columns: layout ID, code type and code.'); ?>
+</p>
+
+<p class='text'>
+<table border='1' cellpadding='4'>
+ <tr bgcolor='#dddddd' class='dehead'>
+  <td><?php echo xlt('Category'); ?></td>
+  <td><?php echo xlt('Layout'); ?></td>
+  <td><?php echo xlt('Code'); ?></td>
+  <td><?php echo xlt('Description'); ?></td>
+ </tr>
+<?php
+$lastcat = '';
+$lastlayout = '';
+
+$res = sqlStatement("SELECT grp_form_id, grp_title, grp_mapping, grp_services FROM layout_group_properties " .
+  "WHERE grp_group_id = '' AND activity = 1 AND grp_services != '' ORDER BY grp_mapping, grp_title, grp_form_id");
+
+while ($row = sqlFetchArray($res)) {
+  // $jobj = json_decode($row['notes'], true);
+  if ($row['grp_services'] == '*') $row['grp_services'] = '';
+  $codes = explode(';', $row['grp_services']);
+  foreach ($codes as $codestring) {
+    echo " <tr>\n";
+
+    echo "  <td class='detail'>";
+    if ($row['grp_mapping'] != $lastcat) {
+      $lastcat = $row['grp_mapping'];
+      echo text($lastcat);
+    }
+    echo "&nbsp;</td>\n";
+
+    echo "  <td class='detail'>";
+    if ($row['grp_form_id'] != $lastlayout) {
+      $lastlayout = $row['grp_form_id'];
+      echo text($row['grp_title']);
+    }
+    echo "&nbsp;</td>\n";
+
+    echo "  <td class='detail'>";
+    echo text($codestring);
+    echo "</td>\n";
+
+    echo "  <td class='detail'>\n";
+    list ($codetype, $code) = explode(':', $codestring);
+    $crow = sqlQuery("SELECT code_text FROM codes WHERE " .
+      "code_type = ? AND code = ? AND active = 1 " .
+      "ORDER BY id LIMIT 1",
```

Comment:
making sure this does not require access to external codes

Reply:
True as noted above.